test-harness/modules/mocker/endpoints/generic.py

```diff
@@ -0,0 +1,276 @@
+        self.context = context
+        http.server.HTTPServer.__init__(self, *args, **kw)
+
+
+class TcpIpHttpEndpoint(Endpoint):
+    """Base class for all endpoints that serve TCP/IP requests
+
+        This class binds together HTTPd server code, http request handler and
+        endpoint context to form a base class for all endpoints that serve
+        TCP/IP traffic.
+    """
+    def __init__(self, handler_class, port, ip=''):
+        """Initialize new TcpIpHttpEndpoint object
+
+        Args:
+            handler_class (obj): a request handler class that will be handling
+                requests received by internal httpd server
+            port (int): tcp port that httpd server will listen on
+            ip (str): ip address that httpd server will listen on, by default
+                listen on all addresses
+        """
+        endpoint_id = "http://{}:{}".format(ip, port)
+        super().__init__(endpoint_id)
+
+        self._context.data['listen_ip'] = ip
+        self._context.data['listen_port'] = port
+
+        self._handler_class = handler_class
+
+        self.__setup_httpd_thread(ip, port)
+
+    def __setup_httpd_thread(self, ip, port):
+        """Setup internal HTTPd server that this endpoints relies on to serve
+           requests.
+        """
+        self._httpd = StatefullHTTPServer(self._context,
+                                          (ip, port),
+                                          self._handler_class)
+
+        httpd_thread_name = "TcpIpHttpdThread-{}".format(self.id)
+        self._httpd_thread = threading.Thread(target=self._httpd.serve_forever,
+                                              name=httpd_thread_name)
+
+
+class UnixSocketStatefulHTTPServer(StatefullHTTPServer):
+    """Base class for all endpoint-internal httpd servers that listen on
+       Unix socket.
+
+    This class inherits from StatefullHTTPServer and mofies it's behaviour
+    so that it's able to listen on Unix socket.
+
+    Attributes:
+        address_family: set only to override default value of the variable set
+            in the http.server.HTTPServer class, must not be modified.
+    """
+    address_family = socket.AF_UNIX
+
+    def server_bind(self):
+        """Override default server socket bind behaviour to adapt it to
+           serving on Unix socket.
+
+        Please check the documentation of http.server.HTTPServer class for more
+        details.
+        """
+        socketserver.TCPServer.server_bind(self)
+        self.server_name = self.context.data['socket_path']
+        self.server_port = 0
+
+    def client_address(self):
+        """Override default client_address method to adapt it to serving on Unix
+        socket. Without it logging will break as Unix socket has no notion of
+        the client's IP address.
+
+        Please check the documentation of http.server.HTTPServer class for more
+        details.
+        """
+        return (self.context.data['socket_path'], 0)
```

Comment:
There is a python FAQ entry: https://docs.python.org/3.6/faq/library.html#what-kinds-of-global-value-mutation-are-thread-safe

The idea is to not to add locking when it is not necessary and thus make the code cleaner. If there is a popular demand for being explicit I am also fine with it.